slsim/Sources/QuasarCatalog/quasar_pop.py

```diff
@@ -0,0 +1,343 @@
+        :param beta: Faint end slope of quasar luminosity density profile.
+        :type beta: float
+        :param phi_star: Renormalization of the quasar luminosity function for a given h.
+        :type phi_star: float
+        :param cosmo: Cosmology object.
+        :type cosmo: ~astropy.cosmology object
+        :param sky_area: Sky area for sampled quasars in [solid angle].
+        :type sky_area: `~Astropy.units.Quantity`
+        :param noise: Poisson-sample the number of galaxies in quasar density lightcone.
+        :type noise: bool
+        :param redshifts: Redshifts for quasar density lightcone to be evaluated at.
+        :type redshifts: np.ndarray
+        """
+        self.h = h
+        self.zeta = zeta
+        self.xi = xi
+        self.z_star = z_star
+        self.alpha = alpha
+        self.beta = beta
+        self.phi_star = phi_star
+        self.cosmo = cosmo if cosmo is not None else FlatLambdaCDM(H0=70, Om0=0.3)
+        self.sky_area = (
+            sky_area if sky_area is not None else Quantity(0.05, unit="deg2")
+        )
+        self.noise = noise
+        self.redshifts = (
+            np.array(redshifts) if redshifts is not None else np.linspace(0.1, 5.0, 100)
+        )
+
+    def M_star(self, z_value):
+        """Calculates the break absolute magnitude of quasars for a given redshift
+        according to Eq. (11) in Oguri & Marshall (2010): DOI:
+        10.1111/j.1365-2966.2010.16639.x.
+
+        :param z_value: Redshift value.
+        :type z_value: float or np.ndarray
+        :return: M_star value.
+        :rtype: float or np.ndarray :unit: mag
+        """
+        z_value = np.atleast_1d(z_value)
+        denominator = (
+            np.sqrt(np.exp(self.xi * z_value)) + np.sqrt(np.exp(self.xi * self.z_star))
+        ) ** 2
+        result = (
+            -20.90
+            + (5 * np.log10(self.h))
+            - (
+                2.5
+                * np.log10(
+                    np.exp(self.zeta * z_value)
+                    * (1 + np.exp(self.xi * self.z_star))
+                    / denominator
+                )
+            )
+        )
+
+        if np.any(denominator == 0):
+            raise ValueError(
+                "Encountered zero denominator in M_star calculation. Check input values."
+            )
+
+        return result
+
+    def dPhi_dM(self, M, z_value):
+        """Calculates dPhi_dM for a given M and redshift according to Eq (10) in Oguri &
+        Marshall (2010): DOI: 10.1111/j.1365-2966.2010.16639.x.
+
+        :param M: Absolute i-band magnitude.
+        :type M: float or numpy.ndarray
+        :param z_value: Redshift value.
+        :type z_value: float or numpy.ndarray
+        :return: dPhi_dM value.
+        :rtype: float or np.ndarray :unit: mag^-1 Mpc^-3
+        """
+        M = np.atleast_1d(M)
+        z_value = np.atleast_1d(z_value)
+
+        if z_value.shape == ():
+            z_value = np.full_like(M, z_value)
```

Comment:
you marked Narayan's comment as resolved while the tests still did not go and execute the lines of code. Maybe I missed it why you marked them as resolved

Reply:
This was before I used the coverage command to check my code coverage without pushing the PR. That's my fault.